test/asg/Main.hs

```diff
@@ -710,6 +713,59 @@ justNothingIntro = runIntroFormTest "justNothingIntro" expectedThunk $ do
     expectedThunk :: ValT AbstractTy
     expectedThunk = ThunkT expectedComp
 
+-- pattern matching
+
+matchMaybe :: TestTree
+matchMaybe = runIntroFormTest "matchMaybe" (BuiltinFlat IntegerT) $ do
+  unit <- AnId <$> lit AUnit
+  scrutinee <- ctor "Maybe" "Just" (Vector.singleton unit) (Vector.singleton Nowhere)
+  nothingHandler <- lazyLam (Comp0 $ ReturnT (BuiltinFlat IntegerT)) (AnId <$> lit (AnInteger 0))
+  justHandler <- lazyLam (Comp0 $ BuiltinFlat UnitT :--:> ReturnT (BuiltinFlat IntegerT)) (AnId <$> lit (AnInteger 1))
+  result <- match (AnId scrutinee) (AnId <$> Vector.fromList [justHandler, nothingHandler])
+  typeIdTest result
+
+matchList :: TestTree
+matchList = runIntroFormTest "matchList" (BuiltinFlat IntegerT) $ do
+  unit <- AnId <$> lit AUnit
+  nilUnit <- ctor "List" "Nil" mempty (Vector.singleton $ There (BuiltinFlat UnitT))
+  scrutinee <- ctor "List" "Cons" (Vector.fromList [unit, AnId nilUnit]) (Vector.singleton Nowhere)
+  let nilHandlerTy = Comp0 $ ReturnT (BuiltinFlat IntegerT)
+      consHandlerTy =
+        Comp0 $
+          BuiltinFlat UnitT
+            :--:> Datatype "List_F" (Vector.fromList [BuiltinFlat UnitT, Datatype "List" (Vector.singleton $ BuiltinFlat UnitT)])
+            :--:> ReturnT (BuiltinFlat IntegerT)
+  nilHandler <- lazyLam nilHandlerTy (AnId <$> lit (AnInteger 0))
+  consHandler <- lazyLam consHandlerTy (AnId <$> lit (AnInteger 0))
+  result <- match (AnId scrutinee) (AnId <$> Vector.fromList [nilHandler, consHandler])
+  typeIdTest result
+
+-- forall a. Maybe a -> List a
+maybeToList :: TestTree
+maybeToList = runIntroFormTest "maybeToList" maybeToListTy $ do
+  thonk <- lazyLam maybeToListCompTy $ do
+    let nothingHandlerTy = Comp0 $ ReturnT (dtype "List" [tyvar (S Z) ix0])
+        justHandlerTy = Comp0 $ tyvar (S Z) ix0 :--:> ReturnT (dtype "List" [tyvar (S Z) ix0])
+    nothingHandler <- lazyLam nothingHandlerTy $ do
+      tvA <- boundTyVar (S Z) ix0
+      AnId <$> ctor "List" "Nil" mempty (Vector.singleton (Here tvA))
+    traceM "nothingHandler"
+    justHandler <- lazyLam justHandlerTy $ do
+      tvA <- boundTyVar (S Z) ix0
+      vA <- AnArg <$> arg Z ix0
+      nil <- AnId <$> ctor "List" "Nil" mempty (Vector.singleton (Here tvA))
+      AnId <$> ctor "List" "Cons" (Vector.fromList [vA, nil]) (Vector.singleton Nowhere)
+    traceM "justHandler"
+    scrutinee <- AnArg <$> arg Z ix0
+    AnId <$> match scrutinee (AnId <$> Vector.fromList [justHandler, nothingHandler])
+  typeIdTest thonk
+  where
+    maybeToListCompTy :: CompT AbstractTy
+    maybeToListCompTy = Comp1 (dtype "Maybe" [tyvar Z ix0] :--:> ReturnT (dtype "List" [tyvar Z ix0]))
+
+    maybeToListTy :: ValT AbstractTy
+    maybeToListTy = ThunkT maybeToListCompTy
```

Comment:
Would help to add explanatory comments (similar to other tests) about what's (conceptually) going on in each test.